### ISSUE TITLE
Is it possible to also have Master templates?

### ISSUE THREAD

**Issue by zam6ak** [jsreport/jsreport-child-templates#3]
Is it possible to have master (aka layout) templates as well?

For example, consider
**master-html-template**

```
<html>
{#child headerTemplate}
<div>
{#mainReport} <-- or whatever helper would go here -->
</div>
{#child footerTemplate}
</html>
```

then:
**report-one-template**

```
{#layout master-html}
<-- here I would iterate over results, or whatever -->
```

**report-two-template**

```
{#layout master-html}
<-- this is a second report using same master layout -->
```

**Comment by pofider** [comment]
This would need to be implemented as another extension...

There is also possible solution using custom script, take a look at this example
https://playground.jsreport.net/studio/workspace/HJXj32DJx/4

``` js
function beforeRender(req, res, done) {
    req.template.content = '{#child header}' +
        req.template.content + '{#child footer}'
    done()
}
```

I've got this feature request already before, however Is this really so useful? It only saves you couple of lines of code. The solution with script is maybe more universal, what you think?

**Comment by zam6ak** [comment]
I think it would be very useful.
In the system where reports are primarily HTML for example, one can have a lot of reports (I mean a lot!)
Having a master layout template with ability to include child templates would 
1) improve consistency
2) make changes trivial

And in case of HTML, it is not just "couple lines of code" - for example a complex Bootstrap template could have a very complex layout.

A custom script seems ok for simple layout (just include header and footer) but anything more complex will end up in sea of string concatenations and escapes :)

Regards
Z!

**Comment by pofider** [comment]
> And in case of HTML, it is not just "couple lines of code" - for example a complex Bootstrap template could have a very complex layout.

Yes, but you can use child templates, so basically it is just two lines. All the complex layout is in header and footer template

``` html
{#child header}
my template....
{#child footer}
```

> A custom script seems ok for simple layout (just include header and footer) but anything more complex will end up in sea of string concatenations and escapes :)

This is basically the same, you still add only child templates references because the script runs before child templates are evaluated. Please check the [playground example](https://playground.jsreport.net/studio/workspace/HJXj32DJx/4) again. The script will be always the same no matter how the layout is complex.

I just want to decide here if it is not better to document the implementation with script than implementing and maintaining a new extension.

**Comment by zam6ak** [comment]
Jan

Sorry, but I feel this is incomplete approach.
Even in simple example you posted,you are essentially concatenating 2 'pieces' of HTML.
With child templates you still have to include header and footer on each reports' template.
Or have this function on each...
In case of the function, if there are a many child templates, ordering would matter and it would get messy...

Just my thought.

**Comment by pofider** [comment]
Thank you for this discussion...

> With child templates you still have to include header and footer on each reports' template.

Yes, so it is two lines (header and footer) vs one line specification of the master template

> In case of the function, if there are a many child templates, ordering would matter and it would get messy...

Please explain this.

**Comment by zam6ak** [comment]
Let's say we want to use something like this for a layout:
https://v4-alpha.getbootstrap.com/examples/dashboard/

Let's suppose we had this as a layout with following pieces
master.hb

```
- header
-- navbar
- left-nav
- content-area
-- chart-section
-- table-section
- footer
```

each 'piece' will be fed additional data via server response (title, nav items, table data, etc, etc)
each piece is self contained, and reusable (e.g. I could use navbar in another layout)

are you saying it is easier to have a function concatenating these pieces for each report
or to have a report that just specifies a layout?

**Comment by pofider** [comment]
My idea was to group these into two "artificial" layout templates

`layout_front` would include header, navbar, left-nav
`layout_end` would include content-area, chart-section, ,table-section, footer

and the script function just adds `layout_front` and `layout_end`. However it is a bit obscure.

**Comment by zam6ak** [comment]
Yes....Ultimately, each of these should be a 'reusable' child that can be added to any layout.
For example, layout-one would have left-nav on left while layout-two could have it on right side...

Have not looked at this in any depth, but it may help...
https://github.com/shannonmoeller/handlebars-layouts
